renku/core/management/migrate.py

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2017-2020 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Renku migrations management.
+
+Migrations files are put in renku/core/management/migrations directory. Name
+of these files has m_1234__name.py format where 1234 is the migration version
+and name can be any alphanumeric and underscore combination. Migration files
+are sorted based on their lowercase name.
+
+When executing a migration, the migration file is imported as a module and all
+functions defined in __all__ are executed sequentially. Migration version is
+checked against the Renku project version (in .renku/metadata.yml) and any
+migration which has a higher version is applied to the project.
+"""
+import importlib
+import re
+from pathlib import Path
+
+import pkg_resources
+
+SUPPORTED_PROJECT_VERSION = 3
+
+
+def is_migration_required(client):
+    """Check if project requires migration."""
+    return (
+        _is_renku_project(client) and
+        _get_project_version(client) < SUPPORTED_PROJECT_VERSION
+    )
+
+
+def is_project_unsupported(client):
+    """Check if this version of Renku cannot work with the project."""
+    return (
+        _is_renku_project(client) and
+        _get_project_version(client) > SUPPORTED_PROJECT_VERSION
+    )
+
+
+def migrate(client):
+    """Apply all migration files to the project."""
+    if not _is_renku_project(client):
+        return
+
+    project_version = _get_project_version(client)
+    migration_executed = False
+
+    for version, path in get_migrations():
+        if version > project_version:
+            module = importlib.import_module(path)
+            for name in module.__all__:
```

Comment:
I think it would be cleaner to define a single entrypoint for each migration file that gets executed (e.g. `def migrate(client):`) instead of relying on `__all__`. With `__all__` you still need to manually add all methods yourself, so it's not less work than putting them inside a function, so I'd go with the simpler interface.